src/program/mod.rs

```diff
@@ -155,11 +164,48 @@ impl Program {
                 },
                 Instruction::Delay(delay) => delay.qubits.clone(),
                 Instruction::Fence(fence) => fence.qubits.clone(),
+                Instruction::Capture(capture) => capture.frame.qubits.clone(),
+                Instruction::Pulse(pulse) => pulse.frame.qubits.clone(),
+                Instruction::RawCapture(raw_capture) => raw_capture.frame.qubits.clone(),
                 _ => vec![],
             })
             .collect::<HashSet<_>>()
     }
 
+    /// Simplify this program into a new [Program] which contains only instructions
+    /// and definitions which are executed; effectively, perform dead code removal.
+    ///
+    /// Removes:
+    /// - All calibrations, following calibration expansion
+    /// - Frame definitions which are not used by any instruction such as `PULSE` or `CAPTURE`
+    /// - Waveform definitions which are not used by any instruction
+    ///
+    /// When a valid program is simplified, it remains valid.
+    pub fn into_simplified(&self) -> Result<Self> {
+        let mut expanded_program = self.expand_calibrations()?;
+        expanded_program.calibrations = CalibrationSet::default();
+
+        let mut frames_used: HashSet<&FrameIdentifier> = HashSet::new();
+        let mut waveforms_used: HashSet<&String> = HashSet::new();
+
+        for instruction in &expanded_program.instructions {
+            if let Some(frames) = expanded_program.get_frames_for_instruction(instruction, false) {
+                frames_used.extend(frames.into_iter())
```

Comment:
This should be sufficient, as `extend` takes an `impl IntoIterator`.

```suggestion
                frames_used.extend(frames)
```